Source/loadsave.cpp

```diff
@@ -1974,6 +1974,11 @@ void LoadGame(bool firstflag)
 	}
 
 	gbIsHellfireSaveGame = gbIsHellfire;
+	if (Quests[Q_DIABLO]._qactive == QUEST_DONE) {
+		PlaySFX(USFX_DIABLOD);
+		gbProcessPlayers = false;
+		myPlayer._pmode = PM_QUIT;

```

Comment:
Is this needed?
I thought `PrepDoEnding` will set this and then the game ends? 🤔

Reply:
the game still ends, but the camera is shaking when the game is loaded. This stops it from shaking.

Reply:
I read the issue. My point was: is the line `myPlayer._pmode = PM_QUIT;` necessary or not.
Sorry if I didn't clarify good enough. 😢 

The program flow is this:
1. In [`DiabloDeath`](https://github.com/diasurgical/devilutionX/blob/a66ca446950bceb25ac31ee268dfc8e2ec9259ce/Source/monster.cpp#L955) `gbProcessPlayers` is set to false, but `PM_QUIT` isn't set.
2. Then the players saves the game.
3. `PM_QUIT` is later set in [`PrepDoEnding`](https://github.com/diasurgical/devilutionX/blob/a66ca446950bceb25ac31ee268dfc8e2ec9259ce/Source/monster.cpp#L4151) (or when starting a new game `GamemenuNewGame`).
4. After this the video starts and the game ends.

If we load the save game, we need to restore the game state to the save game state (as it was in 2.). That means we should only need to set `gbProcessPlayers` to false and don't need to set `PM_QUIT` at all. Because even when loading the save game `PrepDoEnding` later sets `PM_QUIT`.
That's why I raised the question. But perhaps I miss something.

Reply:
If I recall correctly, without setting PM_QUIT at this point, the screen was still shaking on load.

Reply:
> Because even when loading the save game PrepDoEnding later sets PM_QUIT.

Are you sure it will call `PrepDoEnding()` on load? I only see it called in two places, on monster death (`MonsterDeath` in monster.cpp, which already occurred by the point of save) and in `ShowProgress` in interfac.cpp which requires `gbSomebodyWonGameKludge == true` but it sets it to false right after running `PrepDoEnding()`

Reply:
`gbProcessPlayers = false` prevents `ProcessPlayers()` from being called, which is what freezes the character's animation. The pmode doesn't really matter at that point. My testing suggests that `PM_QUIT` is indeed not necessary.

`MonsterDeath()` is the function that gets called in `ProcessMonsters()` when the monster's mmode is `MonsterMode::Death`. Diablo's death animation, which is still playing at time of save/load in this scenario, would indeed lead to `PrepDoEnding()` on load.

Reply:
I could have sworn I did all this without PM_QUIT to start with, and only added it after it didn't work. Let me run it again...

Reply:
Without PM_QUIT being set, the issue (#4135) is not resolved, the screen still shakes (tbf it's more like rubberbands, just using the same language as the issue). Just ran it and confirmed.

Reply:
Video:

https://user-images.githubusercontent.com/6080734/163580265-b08386c7-7237-42df-913c-5112a7f5e11c.mp4

Reply:
I see. This is a discrepancy between `Load Game` from the main menu vs the in-game `Load Game` menu option. In the former, `RunGameLoop()` invokes `StartGame()` which leads to `LoadGame()` and then sets `gbProcessPlayers = true` just a few lines later. `PM_QUIT` is a workaround that forces the player into a weird animation state.

Reply:
Yeah - looking through it a bit more, I see that `PM_QUIT` should not have been set yet (I assume it's something like 7 seconds, in normal speed, after Diablo's death that `PrepDoEnding()` is called?) I'm having trouble following what is actually holding the player and screen in place when Diablo is dead under normal conditions.

Reply:
> I assume it's something like 7 seconds, in normal speed, after Diablo's death that PrepDoEnding() is called?

You got it.

> I'm having trouble following what is actually holding the player and screen in place when Diablo is dead under normal conditions.

The player is being held in place by the fact that `ProcessPlayers()` is not called during Diablo's death animation (the aforementioned `gbProcessPlayers = false`). The fact that the player can't do anything is also what holds the camera in place.

The reason `ProcessPlayers()` causes the camera to rubber band during Diablo's death animation because the player's walking animation updates the camera position based on the player's position.

https://github.com/diasurgical/devilutionX/blob/aab970dcdedf9a9ee900e19d227c223a2ddf1fd5/Source/player.cpp#L704-L707

And `ProcessMonsters()` causes the camera to pan toward Diablo's location at a rate of one tile per game tick.

https://github.com/diasurgical/devilutionX/blob/aab970dcdedf9a9ee900e19d227c223a2ddf1fd5/Source/monster.cpp#L1709-L1720

Reply:
Thanks for the help/insight @StephenCWills and @obligaron. I moved the fix to `RunGameLoop` so the `PM_QUIT` workaround isn't needed.